cms/auth/middleware.py

```diff
@@ -0,0 +1,169 @@
+import logging
+from collections.abc import Mapping
+from typing import TYPE_CHECKING
+
+from django.conf import settings
+from django.contrib.auth import get_user_model, login, logout
+from django.contrib.auth.middleware import AuthenticationMiddleware
+
+from cms.auth.utils import validate_jwt
+
+if TYPE_CHECKING:
+    from django.http import HttpRequest
+
+logger = logging.getLogger(__name__)
+User = get_user_model()
+
+JWT_SESSION_ID_KEY = "jwt_session_id"
+
+
+class ONSAuthMiddleware(AuthenticationMiddleware):
+    """Middleware for handling authentication via JWT tokens from AWS Cognito
+    or the default Django authentication.
+
+    When AWS_COGNITO_LOGIN_ENABLED is active, this middleware extracts tokens from cookies, validates
+    them, checks for any mismatch (client IDs, usernames, or token identifiers), and logs out the user
+    if an issue is encountered.
+    """
+
+    def process_request(self, request: "HttpRequest") -> None:
+        """Process each incoming HTTP request to validate JWT tokens and ensure
+        user authentication.
+        """
+        super().process_request(request)
+
+        if not settings.AWS_COGNITO_LOGIN_ENABLED:
+            self._handle_cognito_disabled(request)
+            return
+
+        # Extract tokens from cookies.
+        access_token = request.COOKIES.get(settings.ACCESS_TOKEN_COOKIE_NAME)
+        id_token = request.COOKIES.get(settings.ID_TOKEN_COOKIE_NAME)
+        if not access_token or not id_token:
+            self._handle_unauthenticated_user(request)
+            return
+
+        # Validate the tokens.
+        access_payload = validate_jwt(access_token, token_type="access")  # noqa: S106
+        id_payload = validate_jwt(id_token, token_type="id")  # noqa: S106
+        if not access_payload or not id_payload:
+            logger.info("Invalid or expired JWT tokens; logging out user.")
+            logout(request)
+            return
+
+        # Token and session integrity checks.
+        is_valid = self._validate_token_session_consistency(
+            request=request, access_payload=access_payload, id_payload=id_payload
+        )
+        if not is_valid:
+            logout(request)
+            return
+
+        # Check if the session is up-to-date.
+        access_jti = access_payload["jti"]
+        id_jti = id_payload["jti"]
+        if request.user.is_authenticated and request.session.get(JWT_SESSION_ID_KEY) == f"{access_jti}{id_jti}":
+            # The session is up-to-date.
+            return
+
+        # Update the session with token identifiers.
+        request.session[JWT_SESSION_ID_KEY] = f"{access_jti}{id_jti}"
```

Comment:
Small suggestion, as we only use the two variables for concatenation:

```suggestion
        jwt_session_key = f"{access_payload['jti']}{id_payload['jti']}"
        if request.user.is_authenticated and request.session.get(JWT_SESSION_ID_KEY) == jwt_session_key:
            # The session is up-to-date.
            return

        # Update the session with token identifiers.
        request.session[JWT_SESSION_ID_KEY] = jwt_session_key
```

Reply:
Addressed in [`19cd39d` (#163)](https://github.com/ONSdigital/dis-wagtail/pull/163/commits/19cd39de350fcc17519fe05f42b38c54af137d7a)